### PR TITLE
Hashed password

### PR DESCRIPTION
**Pull request by inik-the-dom**
This is a second pr, it also contains the changes from my last one so as to avoid conflicts.

I did this one separately because you may want to reject it since it will require re-doing the password portion of the config file (*gasp* breaking changes?!). I changed it to not store the password in plain text, but rather salt + hash it. I'm sure this isn't the 'most secure' approach, but the goal was mainly to just not have it in plain text.

### CODE REVIEW

**Comment by TASagent** [comment]
Thanks for the changes.  I've previously implemented password encryption, and wonder if you wouldn't mind using this C# Implementation.  In essence, Each password is stored as a string that's the Hash + its own salt.

```
using System;
using System.Collections;
using System.Collections.Generic;
using System.Security.Cryptography;
using UnityEngine;

public static class Cryptography
{
    /// <summary>
    /// Thank you stack overflow
    /// https://stackoverflow.com/questions/4181198/how-to-hash-a-password/10402129#10402129
    /// </summary>
    public static string HashPassword(string password)
    {
        byte[] salt;
        byte[] hash;
        byte[] hashBytes;

        using (RNGCryptoServiceProvider rng = new RNGCryptoServiceProvider())
        {
            rng.GetBytes(salt = new byte[16]);
        }

        using (var pbkdf2 = new Rfc2898DeriveBytes(password, salt, 10000))
        {
            hash = pbkdf2.GetBytes(20);
        }

        hashBytes = new byte[36];
        Array.Copy(salt, 0, hashBytes, 0, 16);
        Array.Copy(hash, 0, hashBytes, 16, 20);

        return Convert.ToBase64String(hashBytes);
    }

    /// <summary>
    /// Thank you stack overflow
    /// https://stackoverflow.com/questions/4181198/how-to-hash-a-password/10402129#10402129
    /// </summary>
    public static bool ComparePassword(string password, string passwordHash)
    {
        byte[] hashBytes = Convert.FromBase64String(passwordHash);
        byte[] salt = new byte[16];
        byte[] hash;

        Array.Copy(hashBytes, 0, salt, 0, 16);
        using (Rfc2898DeriveBytes pbkdf2 = new Rfc2898DeriveBytes(password, salt, 10000))
        {
            hash = pbkdf2.GetBytes(20);
        }

        for (int i = 0; i < 20; i++)
        {
            if (hashBytes[i + 16] != hash[i])
            {
                return false;
            }
        }

        return true;
    }

}
```

What are your thoughts on this implementation?

**Comment by inik-the-dom** [comment]
I think I see: the main difference is this enables each password to have it's own salt because the salt will be stored concatenated on the password. I agree this is better. 

_Just thinking outloud for below comment:_

As a todo I think passwords should be moved to the db eventually, but I'll probably do that in a future pr since keeping it in the config allows for easier reset via deleting the json value. Otherwise there would need to be a password reset feature.

**Comment by TASagent** [comment]
I agree that with your thoughts on the storage of the password.  The system as it exists now is kind of temporary anyway, and a more robust Username-Password system would be preferable long-term.  I want the control page to be usable by the mods (though not mandatory, of course).  A means of allowing them to control stuff like Permitting TTS messages, replaying notifications, etc.  _Perhaps_ customizable scoping of their abilities. But I think the easiest way to handle that would be if they could do a chat command like `!reset` and the bot would DM them a link to set/reset their password.  The username could automatically be their twitch username.  We could easily get authentication from this, as well as automatically know their level of access to the bot's systems.  If we make these changes, that would definitely be the appropriate time to move credentials to the database.

**Comment by inik-the-dom** [comment]
I took your suggestions, but had to change a few things. Mainly the hash password method always produced a new random salt, which of course will produce a unique hash each time. So I set it up to take a nullable byte array which will enable us to pass a salt to get a consistent hash for comparison. Second, I had to add a method to parse out the salt from the byte[] which I believe was the first 16 bytes for each password.

**Comment by inik-the-dom** [comment]
I just realized I goofed (sort of). The ComparePassword method actually did everything properly, I just misread it I think. It's redundant to the code that for somereason I decided not to change:

```
            if (adminHash == Admin.Password)
            {
                authString = Admin.AuthString;
                return AuthDegree.Admin;
            }
            else if (privHash == Privileged.Password)
            {
                authString = Privileged.AuthString;
                return AuthDegree.Privileged;
            }
            else if (userHash == User.Password)
            {
                authString = User.AuthString;
                return AuthDegree.User;
            }
```

I'll clean that up and use the proper compare password function.

File: TASagentTwitchBot.Core/Config/BotConfiguration.cs [hunk 1]
```diff
@@ -1,6 +1,9 @@
 ﻿using System;
 using System.Text.Json;
 using System.IO;
+using TASagentTwitchBot.Core.Core;
+using System.Text.Json;
+using System.IO;
```
Comment (TASagent) [inline TASagentTwitchBot.Core/Config/BotConfiguration.cs:6]: I like to keep using statements sorted approximately by scope.  System first, then external modules, then internal modules, then namespaces within the project.  But it also looks like the System.IO and System.Text.Json that were added here were duplicated from lines 2 and 3, so they should just be removed

File: TASagentTwitchBot.Core/Core/Cryptography.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace TASagentTwitchBot.Core.Core
```
Comment (TASagent) [inline TASagentTwitchBot.Core/Core/Cryptography.cs:6]: `TASagentTwitchBot.Core.Core` 🤮 

Haha, this naming snafu came from quirks of history, related to how I'd originally set up the project.  `TASagentTwitchBot.Core` is sufficient.

File: TASagentTwitchBot.Core/Core/Cryptography.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Security.Cryptography;
```
Comment (TASagent) [inline TASagentTwitchBot.Core/Core/Cryptography.cs:11]: I keep the using statements at the top of the file, but also half of these are duplicated from the using statements already there

File: TASagentTwitchBot.Core/Config/BotConfiguration.cs [hunk 4]
```diff
@@ -79,18 +80,18 @@ public class AuthConfiguration
 
         public AuthDegree TryCredentials(string password, out string authString)
         {
-
-            if (password == Admin.Password)
+            Cryptography.ComparePassword(password, Admin.Password);
```
Comment (TASagent) [inline TASagentTwitchBot.Core/Config/BotConfiguration.cs:83]: A redundant compare? I'm guessing this was typed out when adapting the new methods, but then left in. :)

**Review by TASagent — CHANGES_REQUESTED** [review]
Just a few small issues I made comments on.  Looks good, though 👍

**Comment by inik-the-dom** [comment]
All your comments have been addressed. Good points, sorry it took 2 weeks for me to set the time aside to do it. I'll be more careful with merge conflicts in the future.

**Review by TASagent — APPROVED** [review]
Sweet, thanks for the changes, Dom!

**Pull request merged**